cql3/expr/expression.cc

```diff
@@ -863,6 +863,26 @@ expression replace_column_def(const expression& expr, const column_definition* n
         }, expr);
 }
 
+expression replace_token(const expression& expr, const column_definition* new_cdef) {
+    return std::visit(overloaded_functor{
+            [] (bool b){ return expression(b); },
+            [&] (const conjunction& conj) {
+                const auto applied = conj.children | transformed(
+                        std::bind(replace_token, std::placeholders::_1, new_cdef));
+                return expression(conjunction{std::vector(applied.begin(), applied.end())});
+            },
+            [&] (const binary_operator& oper) {
+                return std::visit(overloaded_functor{
+                        [&] (const column_value& col) { return expr; },
+                        [&] (const std::vector<column_value>& cvs) { return expr; },
```

Comment:
Nitpick: it's better to replace the above two lines with a default case, for future-proofing:
```c=+
[&] (const auto& default_case) { return expr; }
```

Reply:
When I was writing this code I had a similar idea but decided against it in order for the code to be more strict about types.

Consider the following (maybe artificial) scenario:

```cpp
struct token {};

using var_t = std::variant<token, int, bool>;

void version1(var_t v) {
    std::visit(overloaded {
        [](token t) { puts("Processing token"); },
        [](int i) { /* Do nothing */ },
        [](bool b) { /* Do nothing */ },
    }, v);
}

void version2(var_t v) {
    std::visit(overloaded {
        [](token t) { puts("Processing token"); },
        [](auto i) { /* Do nothing */ }
    }, v);
}
```

If someone now decided to introduce a new `shiny_new_token` struct replacing `token` in variant:
```cpp
struct token {};

struct shiny_new_token {
    token t;
    int new_field;
};

using var_t = std::variant<shiny_new_token, int, bool>;
```

`version1` rightfully would fail to compile (requiring the programmer to fix it), but `version2` would compile fine, but doing the wrong thing (`shiny_new_token` going into `auto` case and doing nothing).

Of course in the actual code (`replace_token`) this issue would have been caught by unit tests. One disadvantage of this approach is that it might add boilerplate code and make changes harder to do (having to fix all `std::visit`s). And maybe in this case `token` will never be replaced by other type and no new types will require the same action as `token` (so safe to do `auto`).

Reply:
Agree with @avelanarius. Better to annoy someone who has to add a new do-nothing case, than to miss a case where we need to add something.

Reply:
Annoyance and code bloat are rarely a good thing.  I'll bet you $10 that out of the next 50 changes to this function, zero will be shiny_new_token.  Almost every future evolution of `expression` will require touching a dozen functions implemented like this, wasting the coders' and reviewers' time.  I thought we invented computers to prevent just that. :)

"Miss a case" scenario is certainly possible (despite likelihood diminished by good testing coverage), but I judge it overall cheaper than the ongoing cost of maintaining implementations like this one everywhere.